src/repositories/versionable/VersionableRepository.ts

```diff
@@ -0,0 +1,88 @@
+import * as mongoose from 'mongoose';
+import { Document, Query, DocumentQuery } from 'mongoose';
+import * as CircularJSON from 'circular-json';
+
+export default class VersionableRepository<D extends mongoose.Document, M extends mongoose.Model<D>> {
+    protected static generateObjectId() {
+        return String(mongoose.Types.ObjectId());
+    }
+
+    private model: M;
+
+    private JSON;
+
+    constructor(model) {
+        this.model = model;
+        this.JSON = JSON;
+    }
+
+    protected create(data: any): Promise<D> {
+        console.log('UserRepository:: create', data);
+        const id = VersionableRepository.generateObjectId();
+        const model = new this.model ({
+            _id: id,
+            originalId: id,
+            ...data,
+        });
+        return model.save();
+    }
+
+    protected count(query): Query<number> {
+        const finalQuery = { deletedAt: undefined, ...query };
+        return this.model.countDocuments(finalQuery);
+    }
+
+    protected getAll(query, projection, options): DocumentQuery<D[], D> {
+        const finalQuery = { deletedAt: undefined, ...query };
+        return this.model.find(finalQuery, projection, options);
+    }
+
+    protected find(query, projection, options): DocumentQuery<D[], D> {
+        const finalQuery = { deletedAt: undefined, ...query };
+        return this.model.find(finalQuery, projection, options);
+    }
```

Comment:
These two functions are identically the same except the name ( performing the same query) so remove one of those.